campaignion_tracking/package/src/gtm.js

```diff
@@ -0,0 +1,335 @@
+      donation: { currencyCode: null, product: null },
+    }
+    this._context = this.loadFromStorage('context') || defaultContext
+
+    this.printDebug('init')
+
+    if (typeof this.tracker === 'undefined') {
+      this.printDebug('No TrackerManager found. Doing nothing.')
+      return
+    }
+
+    if (typeof this.dataLayer === 'undefined') {
+      this.printDebug('No datalayer found. Doing nothing.')
+    }
+
+    /**
+     * Subscribe to messages of the `donation` tracking channel.
+     *
+     * TODO: validate event
+     * TODO: sanitize data
+     */
+    this.donationSubscription = this.tracker.subscribe('donation', (e) => {
+      this.printDebug('campaignion_tracking_gtm', 'handle_donation', e)
+
+      this.printDebug('campaignion_tracking_gtm', 'handle_event', e.name, e.data, e.context)
+
+      // dispatch to my handlers
+      this.dispatch(e.name, e.data, e.context)
+    })
+  }
+
+  /**
+   * Utility function to print to `console.debug`.
+   *
+   * Print only if debug is set to a truthy value.
+   *
+   * @param  {...any} args arguments to print
+   */
+  printDebug (...args) {
+    if (this.debug) {
+      console.debug('[campaignion_tracking]', '(gtm)', ...args)
+    }
+  }
+
+  saveToStorage (key = 'default', data) {
+    this.tracker.saveToStorage('campaignion_tracking:gtm:', key, data)
+  }
+  loadFromStorage (key = 'default') {
+    return this.tracker.loadFromStorage('campaignion_tracking:gtm:', key)
+  }
+  removeFromStorage (key = 'default') {
+    return this.tracker.removeFromStorage('campaignion_tracking:gtm:', key)
+  }
+
+  /**
+   * Dispatch to my handlers.
+   *
+   * The handler are named according to the `eventName`, prefixed with
+   * `handle_`. The handlers are responsible for the actual sending of valid
+   * events to GTM.
+   *
+   * @param {String} eventName the name of the event
+   * @param {object} eventData data of the event
+   * @param {object} context context of the event, e.g. site title, ...
+   */
+  dispatch (eventName = '', eventData = {}, context = {}) {
+    if (typeof this['handle_' + eventName] === 'function') {
+      this['handle_' + eventName](eventName, eventData, context)
+    } else {
+      this.printDebug('no handler for event name:', eventName)
+    }
+  }
+
+  /**
+   * Hook to allow users to manipulate GTM data.
+   *
+   * Maybe some projects/sites want different values for some GTM fields.
+   * Provide a way to allow for that customizations.
+   *
+   * @param {String} eventName the campaignion_tracking event name
+   * @param {object} gtmData the data which will get sent to GTM data layer
+   * @param {object} context the tracking context
+   */
+  callChangeHook (eventName, gtmData, context) {
+    // maybe also inject tracker
+    if (typeof window['campaignion_tracking_change_msg'] === 'function') {
+      gtmData = window.campaignion_tracking_change_msg('gtm', eventName, gtmData, context)
+    }
+    return gtmData
+  }
+
+  updateContext (context) {
```

Comment:
How is this context different from the tracker independent context? What additional information does it contain?

Reply:
Not a all. It's here mainly because there was not need (yet) to have it in the tracker independent place... :)

So it's missing refactoring...